src/query/storages/parquet/src/parquet_rs/pruning.rs

```diff
@@ -0,0 +1,198 @@
+    }
+
+    /// Prune row groups of a parquet file.
+    ///
+    /// Return the selected row groups' indices in the meta.
+    pub fn prune_row_groups(&self, meta: &ParquetMetaData) -> Result<Vec<usize>> {
+        if !self.prune_row_groups {
+            return Ok((0..meta.num_row_groups()).collect());
+        }
+        match &self.range_pruner {
+            None => Ok((0..meta.num_row_groups()).collect()),
+            Some(pruner) => {
+                // Only if the file has row groups level statistics, we can use them to prune.
+                if meta
+                    .row_groups()
+                    .iter()
+                    .any(|rg| rg.columns().iter().any(|c| c.statistics().is_none()))
+                {
+                    return Ok((0..meta.num_row_groups()).collect());
+                }
+                let row_group_stats = collect_row_group_stats(&self.schema, meta.row_groups())?;
+                let mut selection = Vec::with_capacity(meta.num_row_groups());
+
+                for (i, row_group) in row_group_stats.iter().enumerate() {
+                    if pruner.should_keep(row_group, None) {
+                        selection.push(i);
+                    }
+                }
+
+                Ok(selection)
+            }
+        }
+    }
+
+    /// Prune pages of a parquet file.
+    ///
+    /// Return a vector of [`RowSelection`] to represent rows to read.
+    pub fn prune_pages(
+        &self,
+        meta: &ParquetMetaData,
+        row_groups: &[usize],
+    ) -> Result<Option<RowSelection>> {
+        if !self.prune_pages {
+            return Ok(None);
+        }
+        match &self.range_pruner {
+            None => Ok(None),
+            Some(pruner) => {
+                // Only if the file has page level statistics, we can use them to prune.
+                if meta.column_index().is_none() || meta.offset_index().is_none() {
+                    return Ok(None);
+                }
+                let fields = self.schema.leaf_fields();
+                let all_column_index = meta.column_index().unwrap();
+                let all_offset_index = meta.offset_index().unwrap();
+                assert_eq!(all_column_index.len(), all_offset_index.len());
+                assert_eq!(all_column_index.len(), meta.num_row_groups());
+                let mut selectors = Vec::with_capacity(row_groups.len());
+                for r in row_groups.iter() {
+                    let rg = meta.row_group(*r);
+                    let column_index = &all_column_index[*r];
+                    let offset_index = &all_offset_index[*r];
+                    assert_eq!(column_index.len(), offset_index.len());
+                    //  TODO(parquet): we can only collect statistics of columns we need to eval.
+                    // `sel_of_different_cols` is selectors generated by different columns, merge them at last.
+                    let mut sel_of_different_cols = Vec::with_capacity(column_index.len());
```

Comment:
need to project the column for the page index, it's very overhead now.

Reply:
Ok.